### PR TITLE
Auth service, login flow with password confirmation & password reset

### PR DESCRIPTION
**Pull request by drewfitz10**
## Description

- optimized the api service call to `getAllVenues`. It now occurs once in the App.vue on mount and passed as props to subsequent screen that need the venues. This reduced the calls to the api `/locations` from 4 to 1

- Create auth services with needed api requests. Created Vue mixin to call this globally
- Added validation and error states for all forms
- New views added for the password confirmation & password reset
- added protected routes middleware in router so only the intro, password reset, and account confirmation screen are accessible without being logged in
- 

**NOTE:** There are 3 TS errors in the build that I am not 100% sure on how to clean up and could use a hand on them
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x ] My code follows the code style of this project.
- [x ] My change requires a change to the documentation.
- [x ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/components/map-container/map-container-component.vue [hunk 1]
```diff
@@ -46,7 +51,18 @@ export default class MapContainerComponent extends Vue {
...
-    this.createMap();
+    // if the venues exist create the map
+    // This is used for all page visits after app is loaded and inital call
+    // to the API is finished
+    if (!!this.allVenues.length) {
+      this.createMap();
+    }
+
+    // This will watch the prop allVenues and on change createMap()
+    // This is used on initial page load while the API is still fetching the venues
```
Comment (prowe214) [inline src/components/map-container/map-container-component.vue:62]: i wasn't aware of `$watch`, cool!

File: src/components/onboarding/email/enter-email-component.vue [hunk 2]
```diff
@@ -26,22 +47,84 @@ import IntroStepCounter from '@/components/intro-step-counter/intro-step-counter
...
 export default class EnterEmailComponent extends Vue {
-  public entered = '';
-  public placeholder = 'ILove@DenStartupWeek.com';
+  public email = '';
+  public password = '';
+  public showError = false;
+  public accountCreated = false;
+  public resetPasswordWarning = false;
+  public resetSuccess = false;
+  public $auth: any;
```
Comment (prowe214) [inline src/components/onboarding/email/enter-email-component.vue:60]: this `any` is safe 👍

File: src/components/onboarding/email/enter-email-component.vue [hunk 3]
```diff
@@ -26,22 +47,84 @@ import IntroStepCounter from '@/components/intro-step-counter/intro-step-counter
...
     // and if there's chars before and after dot
-    return /^.+@.+\..+$/.test(this.entered);
+    return /^.+@.+\..+$/.test(this.email);
+  }
+
+  get validPassword() {
+    return this.password.length > 6;
+  }
+
+  get formError() {
```
Comment (prowe214) [inline src/components/onboarding/email/enter-email-component.vue]: probably should be named `formValid`

File: src/components/onboarding/email/enter-email-component.vue [hunk 4]
```diff
@@ -26,22 +47,84 @@ import IntroStepCounter from '@/components/intro-step-counter/intro-step-counter
...
+    }
+  }
+
+  public async forgotPassword() {
+    if (!this.validEmail) {
+      this.resetPasswordWarning = true;
+      return;
+    }
+
+    const result = this.$auth.recovery({
```
Comment (prowe214) [inline src/components/onboarding/email/enter-email-component.vue:96]: glad you took this task, I wasn't aware of `$auth.recovery`

File: src/views/confirmation/confirmation.vue [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  props: {
+    confirmationId: String,
+  },
+})
+export default class Confirmation extends Vue {
+  public password = '';
+  public passwordConfirmation = '';
+  public showError = false;
+  public confirmationId!: string;
+  public $auth: any;
```
Comment (prowe214) [inline src/views/confirmation/confirmation.vue:43]: this `any` is safe 👍

File: src/components/map-container/map-container-component.vue [hunk 6]
```diff
@@ -46,7 +51,18 @@ export default class MapContainerComponent extends Vue {
 
   // Lifecycle hook
   private mounted() {
-    this.createMap();
+    // if the venues exist create the map
+    // This is used for all page visits after app is loaded and inital call
+    // to the API is finished
+    if (!!this.allVenues.length) {
```
Comment (drey-stepanyuk) [inline src/components/map-container/map-container-component.vue]: Why do we need the truthy value from this, can't we just check `this.allVenues.length > 0` ?

File: src/views/recovery/recovery.vue [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  props: {
+    recoveryId: String,
+  },
+})
+export default class Reset extends Vue {
+  public password = '';
+  public passwordConfirmation = '';
+  public showError = false;
+  public recoveryId!: string;
+  public $auth: any;
```
Comment (prowe214) [inline src/views/recovery/recovery.vue:43]: this `any` is also safe 👍

**Review by prowe214 — APPROVED** [review]
This looks great.  Huuuuuge bottle of champagne to you bud 🍾 

thank you

File: src/components/onboarding/email/enter-email-component.vue (hunk 3 above)
Comment (drewfitz10) [reply src/components/onboarding/email/enter-email-component.vue]: update made

**Review by drey-stepanyuk — CHANGES_REQUESTED** [review]
Looks good, just one question and also add margin between the email/password input fields when registering. It's currently smushed together.

File: src/components/map-container/map-container-component.vue (hunk 6 above)
Comment (drewfitz10) [reply src/components/map-container/map-container-component.vue]: We can, both operations do the same thing. Updated it to be `.length > 0`. Also pushed updated the spacing on the inputs

**Review by drey-stepanyuk — APPROVED** [review]

**Comment by drey-stepanyuk** [comment]
Sorry it took a while to approve this, I forgot to do it :(

**Pull request merged**